Replace WorkFile stuff with plain BufFiles.

hlinnaka opened the pull request:
Temporary files have been somewhat inconsistent across different
operations. Some operations used the Greenplum-specific "workfile" API,
while others used the upstream BufFile API directly.

The workfile API provides some extra features: workfiles are visible
in the gp_toolkit views, and you can limit their size with the
gp_workfile_limit_* GUCs. The temporary files that didn't go through
the workfile API were exempt, which is not cool.

To make things consistent, remove the workfile APIs. Use BufFiles
directly everywhere. Re-implement the user-facing view and tracking
the limits, on top of the BufFile API, so that those features are not
lost.

The workfile API also supported compressing the temporary files using
zlib. That feature is lost with this commit. If we still want that, we
should re-implement that directly for BufFiles, as a separate commit
later. (If we decide to re-implement it, we should probably use
zstandard rather than zlib, because it faster, which probably matters
more for temporary files than compression ratio.)

Another feature that this removes, is checksumming temporary files.
That doesn't seem very useful, so we can probably live without it. But
if it's still needed, then that should also be re-implemented on top
of the BufFile API, later.

Discussion: https://groups.google.com/a/greenplum.org/d/msg/gpdb-dev/8Xe9MGor0pM/SjqiOo83BAAJ

Eulerizeit commented:
I'm completely in favor of this, however, I think we need to implement the compression method. It is something that we have customers using. 

I also agree that zstd is a better choice and might make the UX for users who are using compression to be a bit better.

berlin-ab commented on src/backend/optimizer/plan/planner.c:
```diff
@@ -59,8 +59,9 @@
 #include "cdb/cdbvars.h"
-
 #include "storage/lmgr.h"
+#include "utils/guc.h"
```
There are several of these standalone guc.h includes. Is that intentional?

hlinnaka replied:
Yes. These files need guc.h, but they were previously getting git indirectly via execWorkfile.h, which used to #include guc.h. Now that execWorkfile.h is gone, these need to include guc.h directly.

hlinnaka commented:
Per discussion on gpdb-dev, compression is still needed. Added a commit to this PR to re-implement spill file compression, with Zstandard.

hlinnaka commented:
Rebased, and fixed the documentation to reflect changes in the GUCs. @mkiyama, can you have a look at the docs changes, please? I did the bare minimum to make the docs not incorrect, but perhaps some of the things should be worded differently.

Also, the docs now say "If enabled, zstd library must be installed on all hosts in the Greenplum Database cluster". That's not new, I just replaced "zlib" with "zstd", but it seems a bit strange to note that requirement here. Whether you build with zlib/zstd support or not, is a compile-time option. I think we usually write the documentation with the assumption that the user is using the official pivotal binaries, with whatever compile-time options we build them with.

mkiyama commented:
OK, I'll take a look.

thanks for the heads up

mkiyama commented:
The changes to the dita files look fine. 
See my questions for an answer to your question about "having zstd installed".

I have a some questions about zlib and zstd ...most not part of this PR.
sorry for so many questions. If you don't know, that's fine. I am confused by GPDB compression stuff in master.

You are changing only the workfile compression to use only zstd is that correct?
If workfile compression only uses zstd, you can remove that sentence about zstd needs to be installed.

What level is zstd using? I believe there are levels 1-19.

There are other place where GPDB gives a choice of zlib, zstd, and QuickLZ (enterprise only)
Your comment seems to indicate that either zlib or zstd is built for GPDB (not both), is that true?
Or are the compression libraries for zlib and sztd both built and included with GPDB (OSS and enterprise)?

Does either zstd or zlib use OS system libraries? 
In 5.x, zlib compression seems to use some system libraries. That's why the line about having zlib installed.

Do you know if GPDB (master) is using zstd as the "default compression"? (if there is such a thing)

yydzero commented:
Shall we enable compression by default? Guess old behavior enables compression by default.

hlinnaka commented:
On 03/01/2019 11:10, Yandong Yao wrote:
> Shall we enable compression by default? Guess old behavior enables compression by default.

Hmm. It was off by default previously. But zstandard is much faster, so 
it might make sense to enable it by default now. It would be good to 
some performance testing on that.

- Heikki

yydzero commented:
> On 03/01/2019 11:10, Yandong Yao wrote: Shall we enable compression by default? Guess old behavior enables compression by default.
> Hmm. It was off by default previously. But zstandard is much faster, so it might make sense to enable it by default now. It would be good to some performance testing on that. - Heikki

Right, gp_workfile_compress_algorithm is none by default.

hlinnaka commented:
This PR is currently failing in the pipeline, because the pipeline is compiling without libzstd support. I'm waiting for that to change, before this can be pushed.

Or, we could add alternative expected output for when the server has been compiled without libzstd support. That would be a nice thing to do anyway, I suppose, although I'd like to have a full concourse run with this PR, and with zstd support, before pushing anyway.

yydzero commented:
Seems the result is incorrect when compression is enabled:

```
insert into t1 select generate_series, generate_series from generate_series(1, 100000000);
insert into t2 select generate_series, generate_series from generate_series(1, 50000000);

yydzero=# select count(1) from t1 left join t2 on t1.c1 = t2.c1;
   count
-----------
 100000000
(1 row)

yydzero=# set gp_workfile_compression = on;
SET
yydzero=# \timing
Timing is on.
yydzero=# select count(1) from t1 left join t2 on t1.c1 = t2.c1;
  count
----------
 12496308
(1 row)
```

hlinnaka commented:
> Seems the result is incorrect when compression is enabled:

Fixed, thanks!

Looks like the test coverage isn't quite there yet, since this wasn't caught earlier. I'll look into adding a test to check for that case, but at least the bug's been fixed now. (I don't want to add that case verbatim, because it takes a long time to run)

yydzero commented:
> > Seems the result is incorrect when compression is enabled:
> 
> Fixed, thanks!
> 
> Looks like the test coverage isn't quite there yet, since this wasn't caught earlier. I'll look into adding a test to check for that case, but at least the bug's been fixed now. (I don't want to add that case verbatim, because it takes a long time to run)

Verified with fix, thanks!

yydzero commented on src/backend/utils/workfile_manager/workfile_mgr.c:
```diff
@@ -1,520 +1,761 @@
 
-#define WORKFILE_SET_MASK  "XXXXXXXXXX"
+/* GUC: Number of unique entries we can hold in the workfile directory */
+int			gp_workfile_max_entries = 8192;
```
Just curious, does this mean there is 8192 work file at most for each query?  What happens if system needs more work file, such as HashJoin node needs 10K batches?

hlinnaka replied:
It's the max number of workfile *sets*. Each hash join counts as one set, regardless of how many actual files there are in the set. So you shouldn't hit the limit if you have one big hash join, but if you have queries that contain thousands of hash joins, then you might.

Thanks, I'll clarify the comments around that.

hlinnaka commented:
> This PR is currently failing in the pipeline, because the pipeline is compiling without libzstd support. I'm waiting for that to change, before this can be pushed.
> 
> Or, we could add alternative expected output for when the server has been compiled without libzstd support. That would be a nice thing to do anyway, I suppose, although I'd like to have a full concourse run with this PR, and with zstd support, before pushing anyway.

I went ahead and modified the tests so that they pass without libzstd support. I also added a test case for the bug that @yydzero found. I think this is ready to be pushed now, I'll just wait for the PR pipeline to churn through the latest version.

Until the pipeline is changed to compile with libzstd support, the compression code won't actually get any testing in the pipeline. But I don't want to wait for that any longer, as we really should get this in before the GPDB 6 beta release, so that there's time to shake out any possible bugs.

hlinnaka commented:
> +1 to merge.
> 
> Also some simple test by using zstd and zlib for workfile on mac:
> 
> ```
> create table t1(c1 int, c2 int);
> create table t2(c1 int, c2 int);
> insert into t1 select generate_series, generate_series from generate_series(1, 100000000);
> insert into t2 select generate_series, generate_series from generate_series(1, 50000000);
> 
> set gp_workfile_compression = on  OR
> set gp_workfile_compress_algorithm='zlib'
> 
> set statement_mem = 500;
> set work_mem = 64;
> 
> select count(1) from t1 left join t2 on t1.c1 = t2.c1;
> ```
> 
> Performance: It takes 180s on my mac using zlib, while takes 110s if using zstd. This is super!

Nice!

> Memory: it takes about 2.4GB using zlib for one backend, while takes 2.8GB using zstd.

Not nice :-(.

> The number of spill file is 8190 for both case. If disable compression, it takes about 350MB memory, so zlib workfile uses about 2.05GB (268K per workfile), zstd work file uses about 2.45GB (321K per workfile).


I tested this on my laptop, and watched memory usage with 'top'. Looking at the peak virtual memory usage ("VIRT"), rather than physical memory usage ("RES"), it's actually much worse than that.

unpatched gp_workfile_compress_algorithm='none': 450 MB
unpatched gp_workfile_compress_algorithm='zlib': 2.7 GB
patched, gp_workfile_compression=on: 10.7 GB

This was pretty bad even before this patch: If you're running with work_mem='64 kB', and the system allocates 2.7 GB, that's a problem. But this patch makes it much worse.

It's understandable. I think libzstd uses a 512 kB buffer for the compression by default. At the peak, the hash join uses about 2x 8190 files, and 2*8192 * 512 kB is about 8 GB, which is in the right ballpark. I think the default buffer size for zlib is 128 kB, which matches with what we're seeing here.

> It will be interesting to think about how to reduce memory usage as a future enhancement for those who are interested in this issue.

Yeah. I don't think this needs to be a blocker for this PR, since it's a pre-existing issue, even though this makes it worse. And I don't think anyone should set `work_mem` to such a small value in production, anyway.

libzstd has parameters to set the window size (and thus memory usage), but unfortunately they're not exposed in the public API.

We could perhaps allocate a smaller buffer, to hold the uncompressed data, and only initialize the compressor after that buffer fills up. I think all the files in this example are < 64 kB (because work_mem=64 kB), so it's quite pointless to allocate a 512 kB, or even 128 kB, buffer. Or perhaps even write out the data uncompressed, until the file size reaches some threshold. Compression probably isn't helping much when you're dealing with tiny files.
